### PR TITLE
Fallback nodeIds ordering

### PR DESCRIPTION
**Pull request by idegtiarenko**

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java [hunk 1]
```diff
@@ -221,9 +221,7 @@ private void allocateUnassigned() {
                         allocationOrdering.sort(assignmentNodeIds),
                         // TODO consider ignored nodes here too?
                         () -> (shard.primary()
-                            ? allocationOrdering.sort(
-                                allocation.routingNodes().stream().map(RoutingNode::nodeId).collect(Collectors.toSet())
-                            )
+                            ? allocationOrdering.sort(allocation.routingNodes().stream().map(RoutingNode::nodeId).toList())
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:224]: Ensure fallback is not affected by immutable set ordering

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java [hunk 2]
```diff
@@ -194,32 +196,12 @@ private void allocateUnassigned() {
...
-                    final var forcedNodeIds = allocation.deciders().getForcedInitialShardAllocationToNodes(shard, allocation);
-                    if (forcedNodeIds.isEmpty()) {
-                        assignmentNodeIds = assignment.nodeIds();
-                    } else {
-                        assignmentNodeIds = forcedNodeIds.get();
-                        if (logger.isTraceEnabled()) {
-                            logger.trace(
-                                "Shard assignment is ignored as shard [{}] initial allocation forced to {}",
-                                shard.shardId(),
-                                assignment
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:210]: This log statement was lying. This parameter supposed to be ` forcedNodeIds.get()`. Fixed and extracted to a method.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java [hunk 3]
```diff
@@ -252,7 +234,7 @@ private void allocateUnassigned() {
                                     }
                                     continue nextShard;
                                 }
-                                case THROTTLE -> isThrottled[0] = true;
+                                case THROTTLE -> isThrottled.set(true);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java:237]: Hmm I don't think this needs the visibility guarantees of an `AtomicBoolean`, we only use it on this one thread.

**Review by DaveCTurner — APPROVED** [review]
LGTM but I'd rather stick with the `boolean[]`

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/bwc

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/bwc

**Comment by idegtiarenko** [comment]
@elasticsearchmachine please run elasticsearch-ci/part-1

**Pull request merged**